From 461491d5c78a5f6d31d728f851328d47dadbb0c9 Mon Sep 17 00:00:00 2001
From: Nishanth Menon <nm@ti.com>
Date: Tue, 17 May 2016 19:07:14 -0500
Subject: [PATCH 189/347] soc: ti: opp-domain: Do continue to support ganged
 voltage supplies

This patch comes from:
  git://git.ti.com/processor-sdk/processor-sdk-linux.git

When the voltage ranges are ganged, then we need ability to setup the
voltage regulator range to absolute max to allow for another rail to go
beyond the limits of the exact range limits of the current OPP.

We can either update the device tree ranges for the same OR we can
ensure that the driver handles the case appropriately.

Reported-by: Carlos Hernandez <ceh@ti.com>
Signed-off-by: Nishanth Menon <nm@ti.com>
Acked-by: Dave Gerlach <d-gerlach@ti.com>
(cherry picked from commit d2ed6b2ebb0953cc438538769dacc7c6bf0d1e4b)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/soc/ti/ti-opp-domain.c |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/drivers/soc/ti/ti-opp-domain.c b/drivers/soc/ti/ti-opp-domain.c
index 6be9ba4..00eaca2 100644
--- a/drivers/soc/ti/ti-opp-domain.c
+++ b/drivers/soc/ti/ti-opp-domain.c
@@ -247,6 +247,16 @@ static int ti_oppdm_do_transition(struct device *dev,
 			 data->vdd_absolute_max_voltage_uv);
 		uv_max = data->vdd_absolute_max_voltage_uv;
 	}
+	/*
+	 * If we do have an absolute max voltage specified, then we should
+	 * use that voltage instead to allow for cases where the voltage rails
+	 * are ganged (example if we set the max for an opp as 1.12v, and
+	 * the absolute max is 1.5v, for another rail to get 1.25v, it cannot
+	 * be achieved if the regulator is constrainted to max of 1.12v, even
+	 * if it can function at 1.25v
+	 */
+	if (data->vdd_absolute_max_voltage_uv)
+		uv_max = data->vdd_absolute_max_voltage_uv;
 
 	do_abb_first = clk_notifier_flags == PM_OPPDM_VOLT_ABORTRATE ||
 	    clk_notifier_flags == PM_OPPDM_VOLT_POSTRATE;
-- 
1.7.5.4

